### PR TITLE
[BPF] expand cttz, ctlz for i32, i64

### LINKED ISSUE #62252: Using `__builtin_clz` and `__builtin_ctz` in bpf target lead to segfault
When using `__builtin_clz` and `__builtin_ctz` in a simple bpf function (see below), Clang exits with
 139 due to segfault. In contrast,`__builtin_popcount` is correctly compiled with a portable implementation.

Looks like the bpf backend lacks portable implementations for these two builtins, and also doesn't handle the error gracefully by returning early with an informative error message but instead crashes.

Here's the minimal bpf program to reproduce using the C scaffolding in [libbpf-bootstrap](https://github.com/libbpf/libbpf-bootstrap), the accompanying `clz.c` and `Makefile` changes are trivial and omitted:
```C
// content of clz.bpf.c

#include <linux/bpf.h>
#include <bpf/bpf_helpers.h>

SEC("tracepoint/syscalls/sys_enter_execve")

unsigned char clz(unsigned long value) {
    return __builtin_clz(value);
}

int bpf_prog(void *ctx)
{
    unsigned long i = 1;
    bpf_printk("number of leading zeros: %u\n", clz(i));
    return 0;
}

char LICENSE[] SEC("license") = "Dual BSD/GPL";
```

Here's the error message when invoking `make clz`:
```console
  BPF      .output/clz.bpf.o
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.	Program arguments: clang -g -O2 -target bpf -D__TARGET_ARCH_arm64 -I.output -I../../libbpf/include/uapi -I../../vmlinux/arm64/ -idirafter /usr/lib64/clang/14.0.6/include -idirafter /usr/local/include -idirafter /usr/include -c clz.bpf.c -o .output/clz.tmp.bpf.o
1.	<eof> parser at end of file
2.	Code generation
3.	Running pass 'Function Pass Manager' on module 'clz.bpf.c'.
4.	Running pass 'BPF DAG->DAG Pattern Instruction Selection' on function '@clz'
 #0 0x0000ffffb22f8f98 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/lib64/libLLVM-14.so+0xb48f98)
 #1 0x0000ffffb22f70e0 llvm::sys::RunSignalHandlers() (/lib64/libLLVM-14.so+0xb470e0)
 #2 0x0000ffffb22334cc (/lib64/libLLVM-14.so+0xa834cc)
 #3 0x0000ffffb95d37d0 (linux-vdso.so.1+0x7d0)
 #4 0x0000ffffb2979e38 llvm::SelectionDAG::createOperands(llvm::SDNode*, llvm::ArrayRef<llvm::SDValue>) (/lib64/libLLVM-14.so+0x11c9e38)
 #5 0x0000ffffb2999054 llvm::SelectionDAG::getNode(unsigned int, llvm::SDLoc const&, llvm::SDVTList, llvm::ArrayRef<llvm::SDValue>, llvm::SDNodeFlags) (/lib64/libLLVM-14.so+0x11e9054)
 #6 0x0000ffffb439ba3c (/lib64/libLLVM-14.so+0x2beba3c)
 #7 0x0000ffffb28b3b0c (/lib64/libLLVM-14.so+0x1103b0c)
 #8 0x0000ffffb28b34f4 llvm::SelectionDAG::Legalize() (/lib64/libLLVM-14.so+0x11034f4)
 #9 0x0000ffffb29aeb28 llvm::SelectionDAGISel::CodeGenAndEmitDAG() (/lib64/libLLVM-14.so+0x11feb28)
#10 0x0000ffffb29ae2d4 llvm::SelectionDAGISel::SelectAllBasicBlocks(llvm::Function const&) (/lib64/libLLVM-14.so+0x11fe2d4)
#11 0x0000ffffb29ac3a4 llvm::SelectionDAGISel::runOnMachineFunction(llvm::MachineFunction&) (/lib64/libLLVM-14.so+0x11fc3a4)
#12 0x0000ffffb2615fc8 llvm::MachineFunctionPass::runOnFunction(llvm::Function&) (/lib64/libLLVM-14.so+0xe65fc8)
#13 0x0000ffffb2425fe4 llvm::FPPassManager::runOnFunction(llvm::Function&) (/lib64/libLLVM-14.so+0xc75fe4)
#14 0x0000ffffb242b694 llvm::FPPassManager::runOnModule(llvm::Module&) (/lib64/libLLVM-14.so+0xc7b694)
#15 0x0000ffffb2426500 llvm::legacy::PassManagerImpl::run(llvm::Module&) (/lib64/libLLVM-14.so+0xc76500)
#16 0x0000ffffb7c348f4 clang::EmitBackendOutput(clang::DiagnosticsEngine&, clang::HeaderSearchOptions const&, clang::CodeGenOptions const&, clang::TargetOptions const&, clang::LangOptions const&, llvm::StringRef, llvm::Module*, clang::BackendAction, std::unique_ptr<llvm::raw_pwrite_stream, std::default_delete<llvm::raw_pwrite_stream> >) (/lib64/libclang-cpp.so.14+0x17a78f4)
#17 0x0000ffffb7f1ff34 (/lib64/libclang-cpp.so.14+0x1a92f34)
#18 0x0000ffffb6e7b618 clang::ParseAST(clang::Sema&, bool, bool) (/lib64/libclang-cpp.so.14+0x9ee618)
#19 0x0000ffffb8599c40 clang::FrontendAction::Execute() (/lib64/libclang-cpp.so.14+0x210cc40)
#20 0x0000ffffb853ae74 clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) (/lib64/libclang-cpp.so.14+0x20ade74)
#21 0x0000ffffb85fbcf4 clang::ExecuteCompilerInvocation(clang::CompilerInstance*) (/lib64/libclang-cpp.so.14+0x216ecf4)
#22 0x0000aaaab2b331b0 cc1_main(llvm::ArrayRef<char const*>, char const*, void*) (/usr/bin/clang-14+0x131b0)
#23 0x0000aaaab2b2ffe4 (/usr/bin/clang-14+0xffe4)
#24 0x0000ffffb8293b3c (/lib64/libclang-cpp.so.14+0x1e06b3c)
#25 0x0000ffffb22331d4 llvm::CrashRecoveryContext::RunSafely(llvm::function_ref<void ()>) (/lib64/libLLVM-14.so+0xa831d4)
#26 0x0000ffffb829439c (/lib64/libclang-cpp.so.14+0x1e0739c)
#27 0x0000ffffb826e21c clang::driver::Compilation::ExecuteCommand(clang::driver::Command const&, clang::driver::Command const*&) const (/lib64/libclang-cpp.so.14+0x1de121c)
#28 0x0000ffffb826e920 clang::driver::Compilation::ExecuteJobs(clang::driver::JobList const&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*> >&) const (/lib64/libclang-cpp.so.14+0x1de1920)
#29 0x0000ffffb82777b0 clang::driver::Driver::ExecuteCompilation(clang::driver::Compilation&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*> >&) (/lib64/libclang-cpp.so.14+0x1dea7b0)
#30 0x0000aaaab2b2daa0 main (/usr/bin/clang-14+0xdaa0)
#31 0x0000ffffb13fa79c __libc_start_call_main (/lib64/libc.so.6+0x2c79c)
#32 0x0000ffffb13fa86c __libc_start_main@GLIBC_2.17 (/lib64/libc.so.6+0x2c86c)
#33 0x0000aaaab2b2f530 _start (/usr/bin/clang-14+0xf530)
clang-14: error: clang frontend command failed with exit code 139 (use -v to see invocation)
clang version 14.0.6 (Red Hat 14.0.6-4.el9_1)
Target: bpf
Thread model: posix
InstalledDir: /usr/bin
clang-14: note: diagnostic msg: 
********************

PLEASE ATTACH THE FOLLOWING FILES TO THE BUG REPORT:
Preprocessed source(s) and associated run script(s) are located at:
clang-14: note: diagnostic msg: /tmp/clz-de7f1c.c
clang-14: note: diagnostic msg: /tmp/clz-de7f1c.sh
clang-14: note: diagnostic msg: 

********************
make: *** [Makefile:111: .output/clz.bpf.o] Error 139
```

My teammate pointed out this is due to a NULL reference. 

```
(lldb) register read
General Purpose Registers:
       rax = 0x000000000051d0e0
       rbx = 0x00007fffffffc9f0
       rcx = 0x0000000000000002
       rdx = 0x0000000000000000
       rdi = 0x0000000000000000
       rsi = 0x0000000000000005
       rbp = 0x000000000051d100
       rsp = 0x00007fffffffc840
        r8 = 0x00007fffffffc8c0
        r9 = 0x0000000000000000
       r10 = 0x000000000051bf48
       r11 = 0x00000000004d0328
       r12 = 0x0000000000000005
       r13 = 0x0000000000000000
       r14 = 0x00000000004d0260
       r15 = 0x000000000051c2f0
       rip = 0x00007ffff2cf919f  libLLVM-14.so.1`llvm::SelectionDAG::createOperands(llvm::SDNode*, llvm::ArrayRef<llvm::SDValue>) + 303
    rflags = 0x0000000000010202
        cs = 0x0000000000000033
        fs = 0x0000000000000000
        gs = 0x0000000000000000
        ss = 0x000000000000002b
        ds = 0x0000000000000000
        es = 0x0000000000000000

(lldb) disassemble --pc
libLLVM-14.so.1`llvm::SelectionDAG::createOperands:
->  0x7ffff2cf919f <+303>: movq   0x30(%rdx), %rdi
```

[clz-de7f1c.zip](https://github.com/llvm/llvm-project/files/11279808/clz-de7f1c.zip)

### PR DESCRIPTION
**Pull request by inclyc**
Fixes: https://github.com/llvm/llvm-project/issues/62252

Depends on: #73667

### CODE REVIEW

**Comment by inclyc** [comment]
Pinging @yonghong-song

**Comment by inclyc** [comment]
@eddyz87 Could you please take a look? This has been stalled for a while :)

**Comment by eddyz87** [comment]
> @eddyz87 Could you please take a look? This has been stalled for a while :)

Hello, I tried this with simple C test:

```c
unsigned int test(unsigned int v) {
  return __builtin_ctz(v);
  //return __builtin_clz(v);
}
```

<details>
<summary>
The <code>clz</code> part compiles fine, but when <code>ctz</code> is used I still get an assertion, however a different one:
</summary>

```
$ clang  --target=bpf -S -O2 test-clz.c -o -
...
LLVM ERROR: Cannot select: t15: i64 = ConstantPool<[32 x i8] c"\00\01\1C\02\1D\0E\18\03\1E\16\14\0F\19\11\04\08\1F\1B\0D\17\15\13\10\07\1A\0C\12\06\0B\05\0A\09"> 0
In function: test
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.	Program arguments: llc -debug-only=isel --asm-show-inst -mtriple=bpf -mcpu=v3 -filetype=obj -o - test-clz.ll
1.	Running pass 'Function Pass Manager' on module 'test-clz.ll'.
2.	Running pass 'BPF DAG->DAG Pattern Instruction Selection' on function '@test'
 #0 0x000055d4977b1db8 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) /home/eddy/work/llvm-project/llvm/lib/Support/Unix/Signals.inc:723:13
 #1 0x000055d4977afeb0 llvm::sys::RunSignalHandlers() /home/eddy/work/llvm-project/llvm/lib/Support/Signals.cpp:106:18
 #2 0x000055d4977b2588 SignalHandler(int) /home/eddy/work/llvm-project/llvm/lib/Support/Unix/Signals.inc:413:1
 #3 0x00007fdbaa03f190 __restore_rt (/lib64/libc.so.6+0x3f190)
 #4 0x00007fdbaa091dec __pthread_kill_implementation (/lib64/libc.so.6+0x91dec)
 #5 0x00007fdbaa03f0c6 gsignal (/lib64/libc.so.6+0x3f0c6)
 #6 0x00007fdbaa0268d7 abort (/lib64/libc.so.6+0x268d7)
 #7 0x000055d497735145 llvm::report_fatal_error(llvm::Twine const&, bool) /home/eddy/work/llvm-project/llvm/lib/Support/ErrorHandling.cpp:125:5
 #8 0x000055d4975f4b6d llvm::SDNode::getValueType(unsigned int) const /home/eddy/work/llvm-project/llvm/include/llvm/CodeGen/SelectionDAGNodes.h:1007:5
 #9 0x000055d4975f4b6d llvm::SDValue::getValueType() const /home/eddy/work/llvm-project/llvm/include/llvm/CodeGen/SelectionDAGNodes.h:1162:16
#10 0x000055d4975f4b6d llvm::SelectionDAGISel::CannotYetSelect(llvm::SDNode*) /home/eddy/work/llvm-project/llvm/lib/CodeGen/SelectionDAG/SelectionDAGISel.cpp:4232:43
```
</details>

<details>
<summary>
Looking at debug info from llc it looks like <code>cttz_zero_undef</code> is expanded using some kind of a lookup table:
</summary>

```
$ llc -debug-only=isel --asm-show-inst -mtriple=bpf -mcpu=v3 -filetype=asm -o - test-clz.ll
...
Type-legalized selection DAG: %bb.0 'test:entry'
SelectionDAG has 7 nodes:
  t0: ch,glue = EntryToken
      t2: i32,ch = CopyFromReg t0, Register:i32 %0
    t3: i32 = cttz_zero_undef t2
  t5: ch,glue = CopyToReg t0, Register:i32 $w0, t3
  t6: ch = BPFISD::RET_GLUE t5, Register:i32 $w0, t5:1



Legalized selection DAG: %bb.0 'test:entry'
SelectionDAG has 18 nodes:
  t0: ch,glue = EntryToken
  t2: i32,ch = CopyFromReg t0, Register:i32 %0
                t8: i32 = sub Constant:i32<0>, t2
              t9: i32 = and t2, t8
            t11: i32 = mul t9, Constant:i32<125613361>
          t13: i32 = srl t11, Constant:i32<27>
        t14: i64 = sign_extend t13
      t16: i64 = add ConstantPool:i64<[32 x i8] c"\00\01\1C\02\1D\0E\18\03\1E\16\14\0F\19\11\04\08\1F\1B\0D\17\15\13\10\07\1A\0C\12\06\0B\05\0A\09"> 0, t14
    t18: i32,ch = load<(load (s8) from constant-pool), zext from i8> t0, t16, undef:i64
  t5: ch,glue = CopyToReg t0, Register:i32 $w0, t18
  t6: ch = BPFISD::RET_GLUE t5, Register:i32 $w0, t5:1
```
</details>

If there is no way to convince lowering to use some other strategy, and you don't want to spend time on implementing translation for `ConstantPool`, I think it would be fine to leave `ctz` as-is, or just adjust error reporting, so that it clearly says that `ctz` is not supported w/o showing a stack-trace. Wdyt?

**Comment by eddyz87** [comment]
Nevermind, I missed the "depends" part :(
Will retest.

**Comment by eddyz87** [comment]
I tried this with the test below using current kernel master, and all works as expected.

```diff
diff --git a/tools/testing/selftests/bpf/progs/verifier_and.c b/tools/testing/selftests/bpf/progs/verifier_and.c
index e97e518516b6..8a051bd0c886 100644
--- a/tools/testing/selftests/bpf/progs/verifier_and.c
+++ b/tools/testing/selftests/bpf/progs/verifier_and.c
@@ -104,4 +104,14 @@ l0_%=:     r0 = 0;                                         \
        : __clobber_all);
 }
 
+unsigned A[3] = {1u << 31, 1u << 30, 1u << 29};
+
+SEC("socket") __success __retval(0) int clz1(void *ctx) { return __builtin_clz(A[0]); }
+SEC("socket") __success __retval(1) int clz2(void *ctx) { return __builtin_clz(A[1]); }
+SEC("socket") __success __retval(2) int clz3(void *ctx) { return __builtin_clz(A[2]); }
+
+SEC("socket") __success __retval(31) int ctz1(void *ctx) { return __builtin_ctz(A[0]); }
+SEC("socket") __success __retval(30) int ctz2(void *ctx) { return __builtin_ctz(A[1]); }
+SEC("socket") __success __retval(29) int ctz3(void *ctx) { return __builtin_ctz(A[2]); }
+
 char _license[] SEC("license") = "GPL";
```

File: llvm/test/CodeGen/BPF/cttz-ctlz.ll [hunk 1]
```diff
@@ -0,0 +1,304 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 4
+; RUN: llc < %s -march=bpf | FileCheck %s
+
+; test that we can expand CTTZ & CTLZ
+
+declare i32 @llvm.cttz.i32(i32, i1)
+
+define i32 @cttz_i32_zdef(i32 %a) {
+; CHECK-LABEL: cttz_i32_zdef:
```
Comment (eddyz87) [inline llvm/test/CodeGen/BPF/cttz-ctlz.ll:9]: Question, how stable are these expansions?
Previously compiler would just error out, maybe just insert some dummy checks that verify that something is returned from these functions? And we can add a few tests on kernel side that verify runtime result. Wdyt?

Comment (inclyc) [reply llvm/test/CodeGen/BPF/cttz-ctlz.ll:9]: > Question, how stable are these expansions?

It is expanded by common codegen functions, thus might be changed without modifications in BPF backend.

> Previously compiler would just error out, maybe just insert some dummy checks that verify that something is returned from these functions? 

I think for compilers "CodeGen" tests, it is common to assert "Generated Asm" without actually run it. For example, for aarch64 backend developers, a "trick" is to write some code on x86 machine, testing the assembly without aarch64 emulator(e.g. qemu-user). 

So IMHO, for LLVM codegen we can just `CHECK-NEXT`: these asm lines. (They can be automatically generated)

> And we can add a few tests on kernel side that verify runtime result.

Yes, that would be nice if we can test "runtime" behavior.

Comment (eddyz87) [reply llvm/test/CodeGen/BPF/cttz-ctlz.ll:9]: > I think for compilers "CodeGen" tests, it is common to assert "Generated Asm" without actually run it. For example, for aarch64 backend developers, a "trick" is to write some code on x86 machine, testing the assembly without aarch64 emulator(e.g. qemu-user).

Actually, I meant to skip matching the full expansion and just do something like below:

```
; CHECK-LABEL: cttz_i32_zdef:
; CHECK:         r0 =
; CHECK:         exit
```

(As we don't really test the expansion logic, we just test that some expansion is applied).

> So IMHO, for LLVM codegen we can just CHECK-NEXT: these asm lines. (They can be automatically generated)

You mean `utils/update_llc_test_checks`, right?
Well, maybe that's good enough, at-least the tests are easy to adjust if anything changes.

Comment (inclyc) [reply llvm/test/CodeGen/BPF/cttz-ctlz.ll:9]: > (As we don't really test the expansion logic, we just test that some expansion is applied)

Hmm, I'm worried about if we do at tests like

```
; CHECK-LABEL: cttz_i32_zdef:
; CHECK:         r0 =
; CHECK:         exit
```

how can this catch BPF backend changes like `setOperationAction()` to `Custom` and do some custom lowering, as if custom lowering might be incorrect though?

I think we should test implementation details about "how to expand / custom" actually. For example if someone really changes the logic of expansion, I think these tests shall be regenerated and we can see such patches (touched our backend tests)

Comment (eddyz87) [reply llvm/test/CodeGen/BPF/cttz-ctlz.ll:9]: Ok, fair enough.

**Review by eddyz87 — APPROVED** [review]

**Pull request merged**